File: iomerge-server-app/src/main/java/com/github/krzychek/server/network/ConnectionHandlerProxy.java
package com.github.krzychek.server.network;


import com.github.krzychek.server.model.message.Message;
import org.pmw.tinylog.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import javax.annotation.PostConstruct;
import java.io.IOException;
import java.net.Socket;


@Component
public class ConnectionHandlerProxy implements ConnectionHandler {

	private final SingleClientHandlerFactory singleClientHandlerFactory;

	private ConnectionHandler connectionHandler;

	@Autowired
	public ConnectionHandlerProxy(SingleClientHandlerFactory singleClientHandlerFactory) {
		this.singleClientHandlerFactory = singleClientHandlerFactory;
	}

	@Override
	public void sendToClient(Message msg) {
		connectionHandler.sendToClient(msg);
	}

	@PostConstruct
	public void resetConnectionHandler() {
		this.connectionHandler = new NOOPConnectionHandler();
	}

	void connect(Socket clientSocket) throws IOException {
		SingleClientHandler clientHandler = singleClientHandlerFactory.createSingleClientHandler(clientSocket);
		this.connectionHandler = clientHandler;
		clientHandler.startReading();
	}

	private static class NOOPConnectionHandler implements ConnectionHandler {

		private final static String MSG = "Call to NOOPConnectionHandler";

		public void sendToClient(Message ignore) { Logger.debug(MSG); }
	}
}

File: iomerge-server-app/src/main/java/com/github/krzychek/iomerge/server/network/ConnectionHandlerProxy.kt
package com.github.krzychek.iomerge.server.network


import com.github.krzychek.iomerge.server.api.appState.AppStateManager
import com.github.krzychek.iomerge.server.model.MessageProcessor
import com.github.krzychek.iomerge.server.model.message.Message
import java.net.Socket
import javax.inject.Inject
import javax.inject.Singleton


@Singleton class ConnectionHandlerProxy
@Inject constructor(private val messageProcessor: MessageProcessor, private val appStateManager: AppStateManager) : ConnectionHandler {

	private var connectionHandler: ConnectionHandler? = null

	override fun sendToClient(msg: Message) = connectionHandler!!.sendToClient(msg)

	fun connect(clientSocket: Socket) {
		SingleClientHandler(clientSocket, messageProcessor, appStateManager).let {
			connectionHandler = it
			it.startReading()
		}
	}
}
